src/simulators/superoperator/superoperator_state.hpp

```diff
@@ -0,0 +1,477 @@
+
+template <class data_t>
+const stringmap_t<Gates> State<data_t>::gateset_({
+  // Single qubit gates
+  {"id", Gates::id},     // Pauli-Identity gate
+  {"x", Gates::x},       // Pauli-X gate
+  {"y", Gates::y},       // Pauli-Y gate
+  {"z", Gates::z},       // Pauli-Z gate
+  {"s", Gates::s},       // Phase gate (aka sqrt(Z) gate)
+  {"sdg", Gates::sdg},   // Conjugate-transpose of Phase gate
+  {"h", Gates::h},       // Hadamard gate (X + Z / sqrt(2))
+  {"t", Gates::t},       // T-gate (sqrt(S))
+  {"tdg", Gates::tdg},   // Conjguate-transpose of T gate
+  // Waltz Gates
+  {"u1", Gates::u1},     // zero-X90 pulse waltz gate
+  {"u2", Gates::u2},     // single-X90 pulse waltz gate
+  {"u3", Gates::u3},     // two X90 pulse waltz gate
+  {"U", Gates::u3},      // two X90 pulse waltz gate
+  // Two-qubit gates
+  {"CX", Gates::cx},     // Controlled-X gate (CNOT)
+  {"cx", Gates::cx},     // Controlled-X gate (CNOT)
+  {"cz", Gates::cz},     // Controlled-Z gate
+  {"swap", Gates::swap}, // SWAP gate
+  // Three-qubit gates
+  {"ccx", Gates::ccx}    // Controlled-CX gate (Toffoli)
+});
+
+//============================================================================
+// Implementation: Base class method overrides
+//============================================================================
+
+template <class data_t>
+void State<data_t>::apply_ops(const std::vector<Operations::Op> &ops,
+                                  OutputData &data,
+                                  RngEngine &rng) {
+  // Simple loop over vector of input operations
+  for (const auto op: ops) {
+    switch (op.type) {
+      case Operations::OpType::barrier:
+        break;
+      case Operations::OpType::gate:
+        // Note conditionals will always fail since no classical registers
+        if (BaseState::creg_.check_conditional(op))
+          apply_gate(op);
+        break;
+      case Operations::OpType::reset:
+        apply_reset(op.qubits);
+        break;
+      case Operations::OpType::matrix:
+        apply_matrix(op.qubits, op.mats[0]);
+        break;
+      case Operations::OpType::kraus:
+        apply_kraus(op.qubits, op.mats);
+        break;
+      case Operations::OpType::superop:
+        BaseState::qreg_.apply_superop_matrix(op.qubits, Utils::vectorize_matrix(op.mats[0]));
+        break;
+      case Operations::OpType::snapshot:
+        apply_snapshot(op, data);
+        break;
+      default:
+        throw std::invalid_argument("QubitSuperoperator::State::invalid instruction \'" +
+                                    op.name + "\'.");
+    }
+  }
+}
+
+template <class data_t>
+size_t State<data_t>::required_memory_mb(uint_t num_qubits,
+                                 const std::vector<Operations::Op> &ops) {
+  // An n-qubit unitary as 2^4n complex doubles
+  // where each complex double is 16 bytes
+  (void)ops; // avoid unused variable compiler warning
+  size_t shift_mb = std::max<int_t>(0, num_qubits + 4 - 20);
```

Comment:
Why ` - 20` ?
Can we add a comment or better, use a `const` with a name?

Reply:
Shift by -20 converts bytes to MiB